character plunge frames/attack implementation - airborne from xianyun

Charlie-Zheng opened the pull request:
- [x] albedo
- [x] alhaitham
- [x] ayaka
- [x] ayato
- [x] barbara
- [x] bennett
- [x] charlotte
- [x] cyno (out/in q state)
- [x] dehya
- [x] diluc
- [x] dori
- [x] eula
- [x] freminet
- [x] furina
- gaming **not in this pr**
- [x] heizou
- [x] hu tao (out/in e state)
- [x] itto
- [x] kaveh
- [x] kazuha
- [x] keqing
- [x] klee
- [x] lynette
- [x] navia
- [x] noelle
- [x] raiden (out/in q state)
- [x] sucrose
- [x] wriothesley
- [x] xiangling
- [x] xianyun
- [x] xiao (out/in q state)
- [x] yae
- [x] yanfei

Charlie-Zheng committed:
Added Bennett plunge

(cherry picked from commit 01e8d41a0e2146f433dddcf94a606175ac8be7cc)
(cherry picked from commit 5dc3de94ac44f3a69f1f85adbb3e8b1ac137ee7e)

Charlie-Zheng committed:
Addedm Albedo plunge

(cherry picked from commit cebbc50c6d6dd7e50d6ffa15e9f1785292f56d9a)

Charlie-Zheng committed:
Added Alhaitham plunge

(cherry picked from commit 6d5f4be2466820640ccbfd9486e89faed90ce93a)

Charlie-Zheng committed:
Fixed can queue after for albedo and bennett

(cherry picked from commit ae9823759757e7cdad4b4f4ee9001ee213c3f999)

Charlie-Zheng committed:
Added ayato plunge

(cherry picked from commit b0fc3d638baedec14f43f2a9eb3f71d6d4782f37)

Charlie-Zheng committed:
Added alhaitham projection and A1 CBs to Xianyun plunges

(cherry picked from commit 0be51dde3a2a5ded79804ef4bdec5f9850316ac1)

Charlie-Zheng committed:
Added barbara plunge

(cherry picked from commit 9f1b9b5225fc477eac8dca0be78aa8a772fb542b)

Charlie-Zheng committed:
Changed jump source to source from player

(cherry picked from commit 956949591da7f542add1d65366281aa4a38053db)

Charlie-Zheng committed:
Merge remote-tracking branch 'genshinsim/main' into xianyun-dev-other-chars-2

Charlie-Zheng committed:
Fixed barbara plunge incorrect init

(cherry picked from commit 098d309a678e40038e99b0d3b59fcabf09ad9cde)

Charlie-Zheng committed:
Merge remote-tracking branch 'genshinsim/main' into xianyun-dev-other-chars-2

srliao commented on internal/characters/bennett/plunge.go:
```diff
@@ -0,0 +1,164 @@
+		c.plungeCollision(collisionHitmark)
+	}
+
+	poiseDMG := 150.0
```
nit: i assume compiler optimizes this variable away. doesn't look like it gets changed anywhere. should probably get rid of the variable here

srliao commented on internal/characters/albedo/plunge.go:
```diff
@@ -0,0 +1,164 @@
+		c.plungeCollision(collisionHitmark)
+	}
+
+	poiseDMG := 150.0
```
nit: see comment on bennett

srliao commented on internal/characters/albedo/plunge.go:
```diff
@@ -0,0 +1,164 @@
+		CanQueueAfter:   highPlungeFrames[action.ActionDash],
+		State:           action.PlungeAttackState,
+	}, nil
+}
```
there needs to be something at the end of the plunge to reset the airborne flag back to grounded. otherwise the status remains `AirborneXianyun`

this is needed on every single plunge action. strictly speaking i guess even jump should have a reset as well to simulate slow falling and no action taken but i think it's fine to ignore that use case

srliao commented:
In addition to my comment re setting status, every jump action should also probably return an error (and therefore ending the sim) if already airborne (i.e. `jump, jump` should crash the sim). 

This is because if we don't simulate slow falling to ground, then you can't jump again while in mid air

Charlie-Zheng committed:
Merge remote-tracking branch 'genshinsim/main' into xianyun-dev-other-chars-2

Charlie-Zheng committed:
Added plunges set airborne source to grounded

Charlie-Zheng commented:
> In addition to my comment re setting status, every jump action should also probably return an error (and therefore ending the sim) if already airborne (i.e. `jump, jump` should crash the sim).
> 
> This is because if we don't simulate slow falling to ground, then you can't jump again while in mid air

Hmm I thought this would be like globally applied to all characters (no actions except plunge can be done while airborne), and it's up to the individual characters to override

srliao commented:
> In addition to my comment re setting status, every jump action should also probably return an error (and therefore ending the sim) if already airborne (i.e. `jump, jump` should crash the sim).
> 
> This is because if we don't simulate slow falling to ground, then you can't jump again while in mid air

retracted. already handled here. i forgot

https://github.com/genshinsim/gcsim/blob/cb8fccbc4ce4f45c48a55e9db081cc621dd3d538/pkg/core/player/exec.go#L127-L129

Charlie-Zheng committed:
Removed venti airborne check and ad

Charlie-Zheng committed:
Adjusted hitmark to be the same as dash timing for non polearms

Charlie-Zheng committed:
Removed ad from venti high plunge. Shifted low plunge to be before high plunge

Charlie-Zheng committed:
Added ayaka plunge

Charlie-Zheng committed:
fix xiao

Charlie-Zheng committed:
refactored poise and radius to be consts

Charlie-Zheng committed:
Added charlotte plunge

Charlie-Zheng committed:
Added cyno plunge

Charlie-Zheng committed:
Fixed cyno plunge canqueueafter

Charlie-Zheng committed:
Added dehya plunge

Charlie-Zheng committed:
Added Diluc plunge

Charlie-Zheng committed:
Added Dori plunge

Charlie-Zheng committed:
Added Eula plunge

Charlie-Zheng committed:
Added Freminet plunge

Charlie-Zheng committed:
Fix cyno q  plunge being infusable

Charlie-Zheng committed:
Added Furina plunge

Charlie-Zheng committed:
Added Heizou plunge

Charlie-Zheng committed:
Fixed wrong plunge scaling for cyno

Charlie-Zheng committed:
Added Hutao plunge

Charlie-Zheng committed:
Removed jump implementations for xianyun buff

Charlie-Zheng committed:
Added Itto Plunge

Charlie-Zheng committed:
Added kaveh plunge

Charlie-Zheng committed:
fix itto and kaveh missing plunge. Added kazuha plunge

Charlie-Zheng committed:
Added keqing plunge

Charlie-Zheng committed:
Added klee plunge

Charlie-Zheng committed:
Added Lynette plunge

Charlie-Zheng committed:
Added navia plunge

Charlie-Zheng committed:
Fixed barbara windup offset wrong direction

Charlie-Zheng committed:
fix furina windup offset wrong direction

Charlie-Zheng committed:
Fixed Klee windup offset wrong direction

Charlie-Zheng committed:
Cleaned up Hutao plunge duplicated data. Fixed plunge collision check

Charlie-Zheng committed:
added noelle plunge

k0l11 committed:
fix albedo plunge
- frame adjustments
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix alhaitham plunge
- frame adjustments
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix ayaka plunge
- adjust hitmark value representation
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix ayato plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- add error when trying to use plunge while in skill state
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix barbara plunge
- use hitmarks directly in frame arrays
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix bennett plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix charlotte plunge
- fix ca windup and CanQueueAfter
- use hitmarks directly in frame arrays
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix cyno plunge
- adjust hitmark value representation
- fix attack offsets
- fix low plunge in q abil name
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix dehya plunge
- fix low plunge -> jump frames
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix diluc plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix dori plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix eula plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix freminet plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix furina plunge
- frame adjustments
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix heizou plunge
- fix CanQueueAfter
- use hitmarks directly in frame arrays
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix hutao plunge
- fix hutao e not infusing plunge
- adjust hitmark value representation
- fix CanQueueAfter
- fix lp/hp abil names in E
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix itto plunge
- fix lp/hp -> charge frames
- remove c6 comment copied from furina
- remove unnecessary infusion handling
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix kaveh plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix kazuha plunge
- fix skill collision dmg not being infused
- attention: breaking change for kazuha high_plunge collide param
- simplify skill plunge glide error check
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix keqing plunge
- adjust lp -> q frames
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix klee plunge
- adjust lp hitmark
- fix missing lp -> dash frames
- use hitmarks directly in frame arrays
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix lynette plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix navia plunge
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
fix noelle plunge
- fix CanQueueAfter
- remove unnecessary burst status check
- use hitmarks directly in frame arrays
- fix attack offsets
- update documentation
- remove unnecessary char key in error msg

k0l11 committed:
Merge branch 'master' into xianyun-dev-other-chars-2

k0l11 committed:
add raiden plunge

k0l11 committed:
add sucrose plunge

k0l11 committed:
add wriothesley plunge

k0l11 committed:
add xiangling plunge

k0l11 committed:
fix xianyun plunge
- move collision hitmarks to const decl
- adjust legal action docs

k0l11 committed:
add xiao plunge
- refactor to accomodate xianyun

k0l11 committed:
add yae plunge

k0l11 committed:
add yanfei plunge
